test/lib/memtable_snapshot_source.hh

```diff
@@ -81,8 +81,12 @@ public:
         , _compactor(seastar::async([this] () noexcept {
             while (!_closed) {
                 // condition_variable::wait() also allocates memory
-                memory::scoped_critical_alloc_section dfg;
-                _should_compact.wait().get();
+                std::optional<memory::scoped_critical_alloc_section> dfg;
+                auto f = _should_compact.wait();
+                // We need to disable the critical alloc scope while waiting on the signal.
+                dfg.reset();
+                f.get();
+                dfg.emplace();
```

Comment:
Isn't it simpler to move the critical section here? Ah, you want to protect against wait() allocating.

Please add a comment, it's a bit tricky.

Reply:
I will add a comment.

Reply:
Restructure this code in v2.